### PR TITLE
Revising sometimes hints v2

### PR DESCRIPTION
**Pull request by matthewkirby**
This is meant to replace #1243. There were 6 months of merge conflicts so I just copied over the very few changes. On top of the changes from before, this now adds the ability to remove sometimes hints when particular settings are on. This was written specifically for disabling Great Fairy Fountain hints when interior ER is on as there is no way to differentiate them.

Unlike #1243, this is also now ready for review.

### CODE REVIEW

File: HintList.py [hunk 1]
```diff
@@ -325,15 +340,16 @@ def tokens_required_by_settings(world):
...
     'Spirit Temple Silver Gauntlets Chest':                        ("the treasure #sought by Nabooru# is", "upon the #Colossus's right hand# is", ['dungeon', 'sometimes']),
     'Spirit Temple Mirror Shield Chest':                           ("upon the #Colossus's left hand# is", None, ['dungeon', 'sometimes']),
     'Spirit Temple MQ Child Hammer Switch Chest':                  ("a #temporal paradox in the Colossus# yields", "a #temporal paradox in the Spirit Temple# yields", ['dungeon', 'sometimes']),
     'Spirit Temple MQ Symphony Room Chest':                        ("a #symphony in the Colossus# yields", "a #symphony in the Spirit Temple# yields", ['dungeon', 'sometimes']),
     'Spirit Temple MQ GS Symphony Room':                           ("a #spider's symphony in the Colossus# yields", "a #spider's symphony in the Spirit Temple# yields", ['dungeon', 'sometimes']),
-    'Shadow Temple Invisible Floormaster Chest':                   ("shadows in an #invisible maze# guard", None, ['dungeon', 'sometimes']),
+    'Shadow Temple Freestanding Key':                              ("#inside a burning skull# lies", "#giant pot in the Shadow Temple# holds", ['dungeon', 'sometimes']),
```
Comment (mracsys) [inline HintList.py:349]: Should be “a” giant pot

File: HintList.py [hunk 2]
```diff
@@ -325,15 +340,16 @@ def tokens_required_by_settings(world):
...
     'Shadow Temple MQ Bomb Flower Chest':                          ("shadows in an #invisible maze# guard", None, ['dungeon', 'sometimes']),
+    'Ice Cavern Map Chest':                                        ("#winds of ice# surround", "a chest #atop a pillar of ice# contains", ['dungeon', 'sometimes']),
+    'Ganons Castle Shadow Trial Golden Gauntlets Chest':           ("#light in the test of darkness# unveils", "a #like-like in Ganon's Castle guards", ['dungeon', 'sometimes']),
```
Comment (mracsys) [inline HintList.py:352]: Like like hint is missing a # on the end

**Review by mracsys — COMMENTED** [review]
A couple minor comments

**Comment by r0bd0g** [comment]
First I want to say I appreciate this introduction of conditional sometimes and how it's being used only for hints that actually just don't work under those settings. I don't want it to be used for things like "oh this hint isn't very useful if this combination of settings is on" because that would be a lot of work and a lot of debates to get it working in a way people would be happy with. I think it's better just to try and call out locations that are generally likely to be useful hints. But hinted great fairy fountains under indoors ER would just be some nonsense to see, so I think it makes perfect sense to exclude those.

Forest Temple Boss Key Chest - This has an equivalent location in MQ that ought to be sometimes hinted as well. This kind of location you probably do have to hint both versions b/c there's nothing to differentiate them through hint text and giving out mq/vanilla information just b/c one version of it is arbitrarily not hinted is some jank I think we need to stay away from. But really I don't feel like either version of the dungeon needs this location hinted since you usually pass it on the way to the floormaster (or mq redead) room anyway, and you're typically full-clearing at this point still looking for keys. You might go to floormaster first if you sequence break it or have hover boots, maybe that's a little faster routing, but that seems like approximately as large of a detour as BK chest and to do it first might require a trick that some players won't know how to do, so if you insist on hinting one of those checks it makes more sense for it to be floormaster instead. As well I want to add that I don't know that that non-clearer hint is ready for prime-time. (The hint actually makes no sense.) I'd recommend just using the clearer hint text as well or something close to it (for example say "in a deep forest", rather than "in the Forest Temple"), since non-clearer hints is supposed to just be a little colourful moreso than completely cryptic.

Water Temple Central Pillar Chest - The equivalent check for this location in MQ should be hinted. The non-clearer hint calls out "Water Temple" specifically, where we usually refer to it as "vast lake". The way its worded though "in the depths of a vast lake" doesn't make it clear that it's a specific location, so I'd again just call out the central pillar same as with the clearer text, and just sub water temple for vast lake. I also want to call out Water Temple Boss Key Chest as a pretty solid location worth hinting, as it has a lot of the same requirements as central pillar (might be able to sub irons for gold scale, only probably need zl, and need one of hovers or longshot), is also at a dead end, and requires you to collect keys to reach. It doesn't make sense to hint the skulltula below that bombable floor which is overall a shorter detour and has arguably lighter requirements, but leave BK Chest out. So I think central pillar is a solid addition but I don't think BK Chest needs to be removed.

Bottom of the Well Lens of Truth Chest - This is actually a pretty solid hint to see with songs on bosses, good information particularly when you still need to find ZL. It's also a pretty useful hint to see being the only check in the dungeon to require Kokiri Sword. But with the invisible check also being in that room (and who out there actually needs the sword for this), I think it's fine to see it go. But I need to call out Bottom of the Well MQ Compass Chest as the equivalent MQ location needing to also be removed. It's an even better removal in MQ since you don't need ZL to reach it. (The Lens chest in MQ BotW is actually a pretty rough location, but it's in the same spot as the Map chest in vanilla, so I'm not sure how would be best to go about adding a sometimes hint for that.)

Shadow Temple Invisible Floormaster Chest - This is actually an amazingly good hint if FAE is enabled, but that's a pretty niche case. This check is being removed from sometimes and has an equivalent MQ location in the 2nd Dead Hand fight. In the past I have had to argue to keep the MQ variant of this hint around, so I know where are some who already want to let it go. It actually can be super annoying to get that dead hand to spawn, though, but being fair the requirements for some of the other checks in that back area are worse. You can reword the MQ check to reference dead hand and/or the bomb flowers, or remove it entirely. I think I would prefer it to keep parity between the two versions of the dungeon by seeing it removed. (Though honestly I think I'd prefer more to keep both around. I'm not sure why other than that I like the flavour text.)

Shadow Temple Freestanding Key - I recommend calling out 'house of the dead' in the non-clearer text. Not all non-clearer hints call out the dungeon like this, but honestly they probably should. The equivalent MQ location is much more difficult to get into, so if this deserves a hint, the MQ variant should get one too. But the room is different enough that it's difficult to think of some text that applies to both. On the other hand the MQ version of the room doesn't have much special about it to call out at all. The main thing they have in common is the previous silver rupee room. I don't know that the MQ version really needs a hint even though it is worse (I don't think the vanilla version needs one either), but I can at least sort of understand why this is here. I dunno, I'm just rambling at this point. No changes really need to be made to the implementation here, just there's some slight inconsistency with the more difficult MQ check not getting a hint. While we're here and calling out fickle locations, the longshot silver rupees in the huge pit room of MQ shadow has screwed me over many a time.

Ice Cavern Map Chest - I don't understand hinting this as the check is completely irrelevant. You have to already have blue fire with you to skip this room, which almost never happens, and even if you have blue fire it'll be more efficient to use it on King Zora instead, unless King Zora is also hinted or you actually manage to already possess two blue fires, which at that point I'm not sure I want that stuff clogging up all of my bottles for such a small optimization. And besides all that it's such a short detour compared to a lot of other locations that go completely unhinted. I don't think this location should be hinted at all. There is an equivalent MQ location, which is in the same room as the freestanding item so it's even more not worth hinting, and if you wanted to hint one but not the other it'll be a reach to think of some text that would apply unambiguously to this check from vanilla only.

Ganons Castle Shadow Trial Golden Gauntlets Chest - This one has an equivalent MQ location that should be called out with an equivalent text between the two versions. I'm not sure what 'light' from the non-clearer text has to do with this check, I mean you can use fire arrows but they're not necessary. The MQ version doesn't have a like like. I recommend something like "deep in the test of darkness lies", "near the end of the Shadow Trial lies a chest containing" or something along those lines. Yeah, there's another check in there, but it's not as deep.

**Review by Zannick — APPROVED** [review]

**Comment by Zannick** [comment]
Is the feedback mostly that we should add MQ equivalents? Or that we need to reduce the likelihood that one can tell from a hint whether a dungeon is MQ? It's all right from my perspective to merge now and do that separately, given that we want these in 6.1 for races that don't use MQ.

**Comment by r0bd0g** [comment]
There's no way to choose whether you want these changes. While we do support racers, it shouldn't come at the expense of all other users. We shouldn't do an official release with an important feature like hints in this half-broken state. My comments on these changes mainly fall into 3 categories.

MQ Equivalency Issues:
- Forest Temple Boss Key Chest (equivalent MQ location should also be hinted)
- Water Temple Central Pillar Chest (equivalent MQ location should also be hinted)
- Bottom of the Well Lens of Truth Chest (equivalent MQ location should not be hinted)
- Shadow Temple Invisible Floormaster Chest (equivalent MQ location either should not be hinted, or the hint changed to be unambiguous between the two versions)
- Ice Cavern Map Chest (equivalent MQ location should be hinted, though it'll be hard to think of a good way to word it)
- Ganons Castle Shadow Trial Golden Gauntlets Chest (equivalent MQ location should also be hinted)

Issues with the way the hints are worded:
- Forest Temple Boss Key Chest (non-clearer hint is unintelligible)
- Water Temple Central Pillar Chest (non-clearer hint nitpick)
- Shadow Temple Freestanding Key (non-clearer hint nitpick)
- Ganons Castle Shadow Trial Golden Gauntlets Chest (non-clearer hint nitpick)

Questions about which locations should be hinted:
- Forest Temple Boss Key Chest (Between this and the Floormaster chest, Floormaster is strictly worse. It doesn't make sense to hint either of them, honestly, but if one is needed, it should be Floormaster.)
- Water Temple Central Pillar Chest (Water Temple Boss Key Chest is a location with similar requirements that is also deserving of a sometimes hint.)
- Shadow Temple Freestanding Key (Equivalent MQ location is strictly worse and should probably also be hinted with a unique text.)
- Ice Cavern Map Chest (It doesn't make sense to hint this location at all since it's not a particularly long check and you almost always need to get Blue Fire from this room anyway.)

**Comment by matthewkirby** [comment]
Cool. I just hit ctrl+q and lost my entire response.

**Comment by matthewkirby** [comment]
Basically, we made revisions for each of these suggestions.

**Comment by r0bd0g** [comment]
While I don't get irons chest hint either since you're probably mainly in ice for this song, I can't object on any kind of consistency grounds.

The last thing then is mainly nitpicks about the non-clearer hint text of some of the new hints. The ones that remain seem okay (aside: non-clearer hint are not actually supposed to be cryptic even though in the past it's sometimes ended up that way -- they should be mostly clear about where's meant, just with an extra sprinkle of flavour), but mainly I suggest using some version of the dungeon's flavour text name (house of the dead, vast lake, etc) if it's not already in use.

**Comment by matthewkirby** [comment]
Alrighty. I can try to do that. Thinking like same as nonclear text but "house of the dead" instead of "shadow temple" for example?

**Comment by r0bd0g** [comment]
Yeah, though if the end result makes it unclear then don't worry about it if you can't make it work. (For the shadow temple example though the hint I'm thinking of doesn't currently mention a dungeon at all.)

**Comment by matthewkirby** [comment]
Made some changes then to the unclear hints

- Water Central Pillar: beneath a #tall tower under a vast lake# lies
- Shadow Key: a #burning skull in the House of the Dead# holds
- Shadow Trial 2: #deep in the test of darkness# lies, also changed clear hint to say "Ganon's Shadow Trial"
- MQ Shadow Stalfos: Revised unclear and wrote clear hint "near an #empty pedestal within the House of the Dead# lies", "#stalfos in the Shadow Temple# guard"

**Comment by r0bd0g** [comment]
I can't remember if house of the dead is normally capitalized or not? All of those seem good to me though.

File: HintList.py [hunk 3]
```diff
@@ -57,15 +57,21 @@ def getHintGroup(group, world):
         if hint.name in world.always_hints and group == 'always':
             hint.type = 'always'
 
+        conditional_keep, type_append = True, False
```
Comment (Zannick) [inline HintList.py:60]: nit: prefer separate lines for assignment

**Review by Zannick — APPROVED** [review]

File: HintList.py (hunk 3 above)
Comment (matthewkirby) [reply HintList.py]: Fixed this

**Pull request merged**